crates/bevy_ecs/src/world/entity_ref.rs

```diff
@@ -756,6 +824,60 @@ impl<'w> EntityWorldMut<'w> {
         unsafe { self.into_unsafe_entity_cell().get_mut_by_id(component_id) }
     }
 
+    /// Gets the component of the given [`TypedComponentId`] from the entity.
+    ///
+    /// You should prefer [`EntityWorldMut::get`] where possible, this is a niche method
+    /// for accessing dynamic components whose types are known at compile time.
+    pub fn get_dynamic<T: Send + Sync + 'static>(
+        &self,
+        component_id: TypedComponentId<T>,
+    ) -> Option<&'_ T> {
+        EntityRef::from(self).get_dynamic(component_id)
+    }
+
+    /// Gets a mutable reference to the component of the given [`TypedComponentId`] from the entity.
+    ///
+    /// You should prefer [`EntityWorldMut::get_mut`] where possible, this is a niche method
+    /// for accessing dynamic components whose types are known at compile time.
+    pub fn get_dynamic_mut<T: Send + Sync + 'static>(
+        &mut self,
+        component_id: TypedComponentId<T>,
+    ) -> Option<Mut<'_, T>> {
+        // SAFETY: we have mutable access to all components in the World
+        unsafe { self.as_unsafe_entity_cell().get_dynamic_mut(component_id) }
+    }
+
+    /// Consumes `Self` and gets a mutable reference to the component of the given [`TypedComponentId`] from the entity.
+    ///
+    /// You should prefer [`EntityWorldMut::into_mut`] where possible, this is a niche method
+    /// for accessing dynamic components whose types are known at compile time.
+    pub fn into_dynamic_mut<T: Send + Sync + 'static>(
+        self,
+        component_id: TypedComponentId<T>,
+    ) -> Option<Mut<'w, T>> {
+        // SAFETY: We have mutable access to all components of this entity.
+        unsafe { self.into_unsafe_entity_cell().get_dynamic_mut(component_id) }
+    }
+
+    /// Inserts a dynamic component into the entity.
+    ///
+    /// This will overwrite any previous value(s) of the same component type.
+    pub fn insert_dynamic<T: Send + Sync + 'static>(
+        &mut self,
+        component_id: TypedComponentId<T>,
+        component: T,
+    ) -> &mut Self {
+        if self.world_id() != component_id.world_id {
+            panic!("Attempted to insert a dynamic component into the wrong World");
+        }
```

Comment:
Just to be sure I would put a `debug_assert!` after all the world id checks that verifies that the type id of `T` is the same as the one in the `ComponentDescriptor` associated to `component_id` in this `World`.

```rs
debug_assert_eq!(self.components().get_info(component_id.id).and_then(|d| d.type_id()), Some(TypeId::of::<T>()));
```

Reply:
good idea.